internal/api_server/agentserver/server.go

```diff
@@ -44,29 +48,65 @@ func New(
 	store store.Store,
 	ca *crypto.CA,
 	listener net.Listener,
+	tlsConfig *tls.Config,
 	metrics *instrumentation.ApiMetrics,
 ) *AgentServer {
 	return &AgentServer{
-		log:      log,
-		cfg:      cfg,
-		store:    store,
-		ca:       ca,
-		listener: listener,
-		metrics:  metrics,
+		log:        log,
+		cfg:        cfg,
+		store:      store,
+		ca:         ca,
+		listener:   listener,
+		tlsConfig:  tlsConfig,
+		metrics:    metrics,
+		grpcServer: NewAgentGrpcServer(log, cfg),
 	}
 }
 
 func oapiErrorHandler(w http.ResponseWriter, message string, statusCode int) {
 	http.Error(w, fmt.Sprintf("API Error: %s", message), statusCode)
 }
 
+func (s *AgentServer) GetGRPCServer() *AgentGrpcServer {
+	return s.grpcServer
+}
 func (s *AgentServer) Run(ctx context.Context) error {
 	s.log.Println("Initializing Agent-side API server")
+	httpAPIHandler, err := s.prepareHTTPHandler()
+	if err != nil {
+		return err
+	}
+
+	grpcServer := s.grpcServer.PrepareGRPCService()
+
+	handler := grpcMuxHandlerFunc(grpcServer, httpAPIHandler)
+	srv := tlsmiddleware.NewHTTPAgentServerWithTLSContext(handler, s.log, s.cfg.Service.AgentEndpointAddress)
+
+	go func() {
+		<-ctx.Done()
+		s.log.Println("Shutdown signal received:", ctx.Err())
+		ctxTimeout, cancel := context.WithTimeout(context.Background(), gracefulShutdownTimeout)
+		defer cancel()
+
+		srv.SetKeepAlivesEnabled(false)
+		_ = srv.Shutdown(ctxTimeout)
+	}()
+
+	s.log.Printf("Listening on %s...", s.listener.Addr().String())
+	srv.TLSConfig = s.tlsConfig
+	if err := srv.ServeTLS(s.listener, "", ""); err != nil && !errors.Is(err, net.ErrClosed) {
```

Comment:
ServeTLS together with the tcp listener is what made it work, otherwise the server cannot negotiate HTTP2 down well to pass it to the gRPC routes.